File: Cpp/567.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if(s1.length() > s2.length()){
            return false;
        }
        unordered_map<char, int> m;
        for(int i = 0; i < s1.length(); i++){
            m[s1[i]]++;
        }
        unordered_map<char, int> t;
        for(int i = 0; i < s1.length(); i++){
            t[s2[i]]++;
        }
        
        if(s1.length() == s2.length()){
            return m == t;
        }
        for(int i = s1.length(); i < s2.length(); i++){
            if(m == t)return true;
            t[s2[i]]++;
            t[s2[i-s1.length()]]--;
            if(t[s2[i-s1.length()]] == 0){
                t.erase(t.find(s2[i-s1.length()]));
            }
        }
        if(m == t)return true;
        return false;
    }
};

File: Python/567.py
class Solution:

    def checkInclusion(self, s1: str, s2: str) -> bool:
        sctr = Counter(s1)
        ll = len(s1)
        dd = Counter(s2[:len(s1)])
        if dd == sctr:
            return True
        n = len(s1)
        for i in range(n, len(s2)):
            dd[s2[i]] = dd.get(s2[i], 0) + 1
            dd[s2[i-n]] = dd.get(s2[i-n], 0) - 1
            if dd.get(s2[i-n]) == 0:
                dd.pop(s2[i-n])
            if dd == sctr:
                return True
        return False
